### ISSUE TITLE
Hystrix Examplesand a little bit more

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2226]
Hystrix Examples&thinsp;&mdash;&thinsp;and a little bit more<br>
<br>
http://bit.ly/2F3Wfph<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<p>This started as a show case for Hystrix. As more tools have been added it might also look for you as a walk through for different tools. Take a look around and see what might fit your development environment.</p>
<p>&thinsp;&mdash;&thinsp;Alexander Schwartz (<a href="mailto:alexander.schwartz@gmx.net">alexander.schwartz@gmx.net</a>)</p>
</div>
</div>
<div>
<h2>1. Hystrix: Make your application resilient</h2>
<div>
<div>
<h3>1.1. Build your application to handle the failure of external services</h3>
<p>Your application has external runtime dependencies, like an external validating service for data your user has entered, or the central CRM of your company.</p>
<p>What will happen to your application when this external service has a problem?</p>
<p>In the worst case your application will become unavailable as well: The screens of your application will freeze and not respond as the external system doesn&rsquo;t return with a response. Users and IT operations will call you because they think it&rsquo;s your fault and don&rsquo;t suspect the external service.</p>
<p>But what should your application do, when the external service is unavailable?</p>
<div>
<ol>
<li>
<p>Provide timely responses to the users.</p>
</li>
<li>
<p>Give a detailed error message to the users.</p>
</li>
<li>
<p>Provide fall backs when the other service is down.</p>
</li>
<li>
<p>Hooks for IT operations for monitoring to pin-point the problematic service.</p>
</li>
<li>
<p>Prevent overloading of the problematic service to avoid domino effects.</p>
</li>
<li>
<p>Automatic recovery once the problematic service is online again.</p>
</li>
</ol>
</div>
</div>
<div>
<h3>1.2. What Hystrix has built in ready-to-use</h3>
<p>These problems are not new. Netflix has solved them for their video on demand service. In 2012 they made their solution open source.</p>
<p>To deliver the above, Hystrix has built in the following defaults:</p>
<div>
<ol>
<li>
<p>Timeout for every request to an external system (default: 1000 ms)</p>
</li>
<li>
<p>Limit of concurrent requests for external system (default: 10)</p>
</li>
<li>
<p>Circuit breaker to avoid further requests (default: when more than 50% of all requests fail)</p>
</li>
<li>
<p>Retry of a single request after circuit breaker has triggered (default: every 5 seconds)</p>
</li>
<li>
<p>Interfaces to retrieve runtime information on request and aggregate level (there&rsquo;s even a ready-to-use realtime dashboard for it)</p>
</li>
</ol>
</div>
<p>It also shows you where you can place you fall back code.</p>
<p>Providing a detailed error message is then still a task for you.</p>
<p>For more details please have a look at <a href="http://hystrix.github.com">http://hystrix.github.com</a>. The <a href="https://github.com/Netflix/Hystrix/wiki">wiki</a> gives detailed information how to use it and the mechanisms inside.</p>
</div>
<div>
<h3>1.3. Hystrix at work</h3>
<p>To seek Hystrix at work in an example application I&rsquo;ve built a REST application that will be put under load soon. Let&rsquo;s start out with the scenario first.</p>
</div>
</div>
</div>
<div>
<h2>2. Prerequisites for this tutorial</h2>
<div>
<div>
<h3>2.1. Java SDK installed</h3>
<p>In order to run this application you&rsquo;ll need to have Java installed. The minimum recommended version is Java JDK 7. I have performed my tests with Java JDK 8. Please download it from <a href="http://www.oracle.com/technetwork/java/javase/downloads/index.html" target="_blank">Oracle Java SE Download page</a>.</p>
</div>
<div>
<h3>2.2. Internet Connection for additional downloads</h3>
<p>In the process of the turorial you&rsquo;ll need to download additional files from the internet. This includes i.e. Apache Tomcat and Apache Maven. All installations are scripted using small Powershell scripts. Expect a download an additional 100 MB during the tutorial.</p>
</div>
<div>
<h3>2.3. Proxy settings</h3>
<p>When using Maven behind a proxy, please setup your proxy for protocols http and https in <code>tools\maven\apache-maven-3.2.x\conf\settings.xml</code>. You should specify the username and password only when this is necessary in your configuration. Please leave it blank otherwise. Please ensure that <code>localhost</code> is listed in <code>nonProxyHosts</code> otherwise the automatic deployment to the locally installed tomcat might not work.</p>
<div>
<p>settings.xml</p>
<div>
<pre>
<code data-lang="xml"><span>&lt;proxy&gt;</span>
  <span>&lt;id&gt;</span>optional<span>&lt;/id&gt;</span>
  <span>&lt;active&gt;</span>true<span>&lt;/active&gt;</span>
  <span>&lt;protocol&gt;</span>http<span>&lt;/protocol&gt;</span>
  <span>&lt;username&gt;</span>proxyuser<span>&lt;/username&gt;</span>
  <span>&lt;password&gt;</span>proxypass<span>&lt;/password&gt;</span>
  <span>&lt;host&gt;</span>proxy.host.net<span>&lt;/host&gt;</span>
  <span>&lt;port&gt;</span>80<span>&lt;/port&gt;</span>
  <span>&lt;nonProxyHosts&gt;</span>localhost|local.net|some.host.com<span>&lt;/nonProxyHosts&gt;</span>
<span>&lt;/proxy&gt;</span>
<span>&lt;proxy&gt;</span>
  <span>&lt;id&gt;</span>optional<span>&lt;/id&gt;</span>
  <span>&lt;active&gt;</span>true<span>&lt;/active&gt;</span>
  <span>&lt;protocol&gt;</span>http<span>&lt;/protocol&gt;</span>
  <span>&lt;username&gt;</span>proxyuser<span>&lt;/username&gt;</span>
  <span>&lt;password&gt;</span>proxypass<span>&lt;/password&gt;</span>
  <span>&lt;host&gt;</span>proxy.host.net<span>&lt;/host&gt;</span>
  <span>&lt;port&gt;</span>80<span>&lt;/port&gt;</span>
  <span>&lt;nonProxyHosts&gt;</span>localhost|local.net|some.host.com<span>&lt;/nonProxyHosts&gt;</span>
<span>&lt;/proxy&gt;</span></code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<h2>3. Scenario to challenge Hystrix: External Bank validation</h2>
<div>
<div>
<h3>3.1. IBAN and BIC validations</h3>
<p>Most of Europe have changed to SEPA in early 2014. Instead of national account number and bank sorting code you now have a <a href="http://en.wikipedia.org/wiki/IBAN">IBAN</a> (International Bank Account Number) and <a href="http://en.wikipedia.org/wiki/ISO_9362">BIC</a> (Bank Identifier Code).</p>
<p>When customers enter their bank details on your website, you want to validate these. As the IBAN includes a check digit, you can implement this check. The BIC doesn&rsquo;t have a check digit, so you can only verify it by lookup in a dictionary.</p>
<p>Things get messy as the BIC might not match the IBAN, or the IBAN might not be valid for direct debit/clearing.</p>
<p>The good news: There are external services that do the validation for you.</p>
</div>
<div>
<h3>3.2. Simulating an external IBAN/BIC validation service</h3>
<p>As we want to show what Hystrix can do for us, we need a IBAN/BIC validation service. We simulate the behaviour with a small Java procedure:</p>
<div>
<p>IBANValidator.java</p>
<div>
<pre>
<code data-lang="java"><span>public</span> <span>class</span> <span>IBANValidator</span> {

    <span>private</span> <span>static</span> <span>Logger</span> LOG = LoggerFactory.getLogger(IBANValidator.class);

    <span>private</span> <span>static</span> DynamicLongProperty timeToWait = DynamicPropertyFactory
            .getInstance().getLongProperty(<span><span>"</span><span>hystrixdemo.sleep</span><span>"</span></span>, <span>100</span>);

    <span>public</span> <span>static</span> <span>synchronized</span> <span>boolean</span> isValid(Account account)
            <span>throws</span> <span>InterruptedException</span> {
        <span>if</span> (<span>Thread</span>.currentThread().isInterrupted()) {
            <span>throw</span> <span>new</span> <span>InterruptedException</span>();
        }
        <span>long</span> t = timeToWait.get();
        LOG.info(<span><span>"</span><span>waiting {} ms</span><span>"</span></span>, t);
        <span>if</span> (t &gt; <span>0</span>) {
            <span>Thread</span>.sleep(t);
        }
        <span>return</span> <span>true</span>;
    }
}</code>
</pre>
</div>
</div>
<p>The core of the simulation is the static <code>isValid()</code> method. As it is synchronized, it will validate one account at a time.</p>
<p>To simulate the processing, we use <code>Thread.sleep()</code>. The delay is by default 100 ms. This can be configured at runtime using an <code>DynamicLongProperty</code>. See the chapter <a href="#archaius">Using runtime configuration</a> below. Changing the delay at runtime will help us later to try out different scenarios.</p>
<p>When someone interrupts the thread while waiting for the <code>synchronized</code> lock or for <code>Thread.sleep()</code>, the method will throw an <code>InterruptedException</code>.</p>
<p>This is a standard Java feature that will be used by Hystrix to notify a service call that is taking too long.</p>
<p>You find this source code in the <code>hystrix-application</code> folder.</p>
</div>
</div>
</div>
<div>
<h2>4. Writing a real application with JAX-RS</h2>
<div>
<div>
<h3>4.1. JAX-RS as a standard for enterprise RESTful web application</h3>
<p>JAX-RS is part of the JavaEE Standard. It defines how a JavaEE application needs to be structured to respond to HTTP-REST-Calls.</p>
<p>REST calls have the following advantages:</p>
<div>
<ol>
<li>
<p>With their JSON-interface they can be interfaced easily from JavaScript clients.</p>
</li>
<li>
<p>The REST API can also be easily tested for functional and load testing (see <a href="#jmeter">Putting your application under load</a> below).</p>
</li>
</ol>
</div>
</div>
<div>
<h3>4.2. RESTeasy as portable JAX-RS implementation</h3>
<p><a href="http://resteasy.jboss.org/">RESTEasy</a> by JBoss is a portable implementation of the standard. It adds also several enterprise features like authentication and signatures.</p>
<p>As a portable implementation I can easily deploy it to a Tomcat application server later.</p>
</div>
<div>
<h3>4.3. Overview of classes to implement a JAX-RS service</h3>
<p>You find the example application in the folder <code>hystrix-application</code>.</p>
<div>
<h4>4.3.1. Application as the starting point</h4>
<p>JAX-RS requires you to name all classes related to the REST part of your application. This class needs to extend <code>javax.ws.rs.core.Application</code>:</p>
<div>
<p>HystrixApplication.java</p>
<div>
<pre>
<code data-lang="java"><span>@ApplicationPath</span>(<span><span>"</span><span>/api</span><span>"</span></span>)
<span>public</span> <span>class</span> <span>HystrixApplication</span> <span>extends</span> Application {
    <span>private</span> <span>Set</span>&lt;<span>Object</span>&gt; singletons = <span>new</span> <span>HashSet</span>&lt;<span>Object</span>&gt;();

    <span>public</span> HystrixApplication() {
        singletons.add(<span>new</span> SimpleSaveAccount());
        singletons.add(<span>new</span> HystrixSaveAccount());
        singletons.add(<span>new</span> ValidationExceptionMapper());
        singletons.add(<span>new</span> InterruptedExceptionMapper());
    }

    <span>@Override</span>
    <span>public</span> <span>Set</span>&lt;<span>Object</span>&gt; getSingletons() {
        <span>return</span> singletons;
    }
}</code>
</pre>
</div>
</div>
<p>This very basic setup lists the two REST endpoints <code>SimpleSaveAccount</code> and <code>HystrixSaveAccount</code> we will look at later plus two exception handlers <code>ValidationExceptionMapper</code> and <code>InterruptedExceptionMapper</code>.</p>
<p>It also defines that the URL <code>/api</code> will be used for all JAX-RS requests relative to the application.</p>
</div>
<div>
<h4>4.3.2. SimpleSaveAccount as REST endpoint</h4>
<p>JAX-RS requires you to name all classes related to the REST part of your application. This class needs to extend <code>javax.ws.rs.core.Application</code>:</p>
<div>
<p>SimpleSaveAccount.java</p>
<div>
<pre>
<code data-lang="java"><span>@Path</span>(<span><span>"</span><span>/simple</span><span>"</span></span>)
<span>public</span> <span>class</span> <span>SimpleSaveAccount</span> <span>extends</span> AbstractSaveAccount {

    <span>@GET</span>
    <span>@Produces</span>(<span><span>"</span><span>text/plain</span><span>"</span></span>)
    <span>public</span> Response hello() {
        <span>return</span> Response.status(Status.OK).entity(<span><span>"</span><span>Hello world</span><span>"</span></span>).build();
    }

    <span>@POST</span>
    <span>public</span> Response save(Account account) <span>throws</span> ValidationException,
            <span>InterruptedException</span> {
        <span>if</span> (!IBANValidator.isValid(account)) {
            <span>throw</span> <span>new</span> ValidationException(<span><span>"</span><span>invalid</span><span>"</span></span>);
        }
        <span>super</span>.saveToDatabase(account);
        <span>return</span> Response.status(Status.OK).build();
    }
}</code>
</pre>
</div>
</div>
<p>This defines for the URL <code>/simple</code> relative to the application defined above how to react on <code>GET</code> and <code>POST</code> requests.</p>
<p>A <code>GET</code>-Request will return with an empty <code>200 OK</code> response. This enables us to use a simple browser request to test the successful deployment of our application.</p>
<p>A <code>POST</code>-Request will have a parameter of type <code>Account</code>. If the account is invalid, it will throw an exception. Otherwise it will save the account. In our sample application the save operation is a dummy implementation that does nothing. After the save returned, the response is <code>200 OK</code>.</p>
</div>
<div>
<h4>4.3.3. ExceptionMappers to handle Exceptions</h4>
<p>Unhandled exceptions result in the application server to present a standard <code>500</code> error page that (depending on the application servers&rsquo;s configuration) might also include a stack trace of the application.</p>
<p>To avoid this and to return REST-ful error messages exceptions are mapped by an <code>ExceptionMapper</code>.</p>
<div>
<p>ValidationExceptionMapper.java</p>
<div>
<pre>
<code data-lang="java"><span>@Provider</span>
<span>public</span> <span>class</span> <span>ValidationExceptionMapper</span> <span>implements</span>
        ExceptionMapper&lt;ValidationException&gt; {

    <span>@Context</span>
    <span>private</span> HttpHeaders headers;

    <span>@Override</span>
    <span>public</span> Response toResponse(ValidationException e) {
        <span>return</span> Response.status(Status.BAD_REQUEST)
                .entity(<span>new</span> Message(e.getMessage()))
                .type(headers.getMediaType()).build();
    }

}</code>
</pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<h2>5. Equip your application with Hystrix</h2>
<div>
<div>
<h3>5.1. Ensuring a clean start up and shut down of Hystrix</h3>
<p>When you want to run code at the start and at the end of a JEE Web application, the place for this is <code>@WebListener</code>. The servlet container will scan for classes with this annotation and run their <code>contextInitialized()</code> and <code>contextDestroyed()</code> methods.</p>
<p>To start up Hystrix there is nothing to do; you can just go ahead and use it.</p>
<p>I recommend a shut down of Hystrix as it starts several thread pools to do its job. Just add the line</p>
<div>
<p>HystrixSetupListener.java</p>
<div>
<pre>
<code data-lang="java">
Hystrix.reset(<span>1</span>, <span>TimeUnit</span>.SECONDS);</code>
</pre>
</div>
</div>
<p>As Hystrix uses Archaius as a default for runtime configuration, we should shut down it as well.</p>
<div>
<p>HystrixSetupListener.java</p>
<div>
<pre>
<code data-lang="java">

<span>if</span> (ConfigurationManager.getConfigInstance() <span>instanceof</span> DynamicConfiguration) {
    ((DynamicConfiguration) ConfigurationManager.getConfigInstance())
            .stopLoading();
} <span>else</span> <span>if</span> (ConfigurationManager.getConfigInstance() <span>instanceof</span> ConcurrentCompositeConfiguration) {
    ConcurrentCompositeConfiguration config =
            ((ConcurrentCompositeConfiguration) ConfigurationManager
                    .getConfigInstance());
    <span>for</span> (AbstractConfiguration innerConfig : config.getConfigurations()) {
        <span>if</span> (innerConfig <span>instanceof</span> DynamicConfiguration) {
            ((DynamicConfiguration) innerConfig).stopLoading();
        }
    }
}</code>
</pre>
</div>
</div>
</div>
<div>
<h3>5.2. Wrap external interfaces as a Hystrix Command</h3>
<p>The core of the application is now set up. To benefit from Hystrix each call needs to be wrapped as a Hystrix command.</p>
<div>
<p>HystrixSaveAccount.java</p>
<div>
<pre>
<code data-lang="java"><span>private</span> <span>static</span> <span>class</span> <span>IBANValidatorCommand</span> <span>extends</span> HystrixCommand&lt;<span>Boolean</span>&gt; {
    <span>private</span> Account account;

    <span>protected</span> IBANValidatorCommand(Account account) {
        <span>super</span>(Setter.withGroupKey(HystrixCommandGroupKey.Factory
                .asKey(<span><span>"</span><span>iban</span><span>"</span></span>)));
        <span>this</span>.account = account; <b>(2)</b>
    }

    <span>@Override</span>
    <span>protected</span> <span>Boolean</span> run() <span>throws</span> <span>Exception</span> {
        <span>return</span> IBANValidator.isValid(account); <b>(3)</b>
    }

}

<span>@POST</span>
<span>@ApiOperation</span>(<span><span>"</span><span>save account data</span><span>"</span></span>)
<span>public</span> Response save(Account account) <span>throws</span> ValidationException,
        <span>InterruptedException</span> {
    <span>try</span> {
        <span>if</span> (!<span>new</span> IBANValidatorCommand(account).execute()) { <b>(1)</b>
            <span>throw</span> <span>new</span> ValidationException(<span><span>"</span><span>invalid</span><span>"</span></span>);
        }
    } <span>catch</span> (HystrixRuntimeException e) { <b>(4)</b>
        <span>if</span> (e.getCause() <span>instanceof</span> <span>InterruptedException</span>) {
            <span>throw</span> (<span>InterruptedException</span>) e.getCause();
        }
        LOG.error(<span><span>"</span><span>problem with command: {}</span><span>"</span></span>, e.getMessage());
        <span>return</span> Response.status(Status.SERVICE_UNAVAILABLE).build();
    }
    <span>super</span>.saveToDatabase(account);
    <span>return</span> Response.status(Status.OK).build();
}</code>
</pre>
</div>
</div>
<div>
<table>
<tbody>
<tr>
<td><b>1</b></td>
<td>constructing the wrapper</td>
</tr>
<tr>
<td><b>2</b></td>
<td>memorizing the parameters</td>
</tr>
<tr>
<td><b>3</b></td>
<td>calling the original code</td>
</tr>
<tr>
<td><b>4</b></td>
<td>taking care of the additional runtime exception that originates from an exception of the services, or from the resilience functionality of Hystrix</td>
</tr>
</tbody>
</table>
</div>
<p>When there is a problem with the Hystrix command there will be a <code>HystrixRuntimeException</code>. This will occur in two cases:</p>
<div>
<ol>
<li>
<p>Hystrix circuit breaker or time out handler are activated.</p>
</li>
<li>
<p>The original command throws an exception.</p>
</li>
</ol>
</div>
<div>
<table>
<tbody>
<tr>
<td></td>
<td>If you want to handle the exceptions of your service as before, you will need to unwrap them.</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div>
<h2>6. Live API documentation</h2>
<div>
<div>
<h3>6.1. Swagger creates a live API at runtime</h3>
<p><a href="http://swagger.io/">Swagger</a> allows you to annotate your REST endpoints. The information from the JAX-RS annotation and the Swagger annotations is collected at runtime.</p>
</div>
<div>
<h3>6.2. Adding Swagger</h3>
<p>To add Swagger to your project, annotate your REST endpoints with <code>@ApiOperation</code> and <code>@Api</code> annotations.</p>
<p>Add the following dependency to your Maven configuration:</p>
<div>
<p>pom.xml</p>
<div>
<pre>
<code data-lang="xml"><span>&lt;dependency&gt;</span>
    <span>&lt;groupId&gt;</span>com.wordnik<span>&lt;/groupId&gt;</span>
    <span>&lt;artifactId&gt;</span>swagger-jaxrs_2.10<span>&lt;/artifactId&gt;</span>
    <span>&lt;scope&gt;</span>compile<span>&lt;/scope&gt;</span>
    <span>&lt;version&gt;</span>1.3.10<span>&lt;/version&gt;</span>
<span>&lt;/dependency&gt;</span></code>
</pre>
</div>
</div>
<p>The following snipped is needed in your web configuration:</p>
<div>
<p>web.xml</p>
<div>
<pre>
<code data-lang="xml"><span>&lt;context-param&gt;</span>
        <span>&lt;param-name&gt;</span>resteasy.resources<span>&lt;/param-name&gt;</span>
        <span>&lt;param-value&gt;</span>com.wordnik.swagger.jaxrs.listing.ApiListingResourceJSON<span>&lt;/param-value&gt;</span>
<span>&lt;/context-param&gt;</span>
<span>&lt;context-param&gt;</span>
        <span>&lt;param-name&gt;</span>resteasy.providers<span>&lt;/param-name&gt;</span>
        <span>&lt;param-value&gt;</span>com.wordnik.swagger.jaxrs.json.JacksonJsonProvider,com.wordnik.swagger.jaxrs.listing.ApiDeclarationProvider,com.wordnik.swagger.jaxrs.listing.ResourceListingProvider<span>&lt;/param-value&gt;</span>
<span>&lt;/context-param&gt;</span>
<span>&lt;servlet&gt;</span>
        <span>&lt;servlet-name&gt;</span>DefaultJaxrsConfig<span>&lt;/servlet-name&gt;</span>
        <span>&lt;servlet-class&gt;</span>com.wordnik.swagger.jaxrs.config.DefaultJaxrsConfig<span>&lt;/servlet-class&gt;</span>
        <span>&lt;init-param&gt;</span>
                <span>&lt;param-name&gt;</span>api.version<span>&lt;/param-name&gt;</span>
                <span>&lt;param-value&gt;</span>1.0.0<span>&lt;/param-value&gt;</span>
        <span>&lt;/init-param&gt;</span>
        <span>&lt;init-param&gt;</span>
                <span>&lt;param-name&gt;</span>swagger.api.basepath<span>&lt;/param-name&gt;</span>
                <span>&lt;param-value&gt;</span>http://localhost:8080/hystrixapp/api<span>&lt;/param-value&gt;</span>
        <span>&lt;/init-param&gt;</span>
        <span>&lt;load-on-startup&gt;</span>2<span>&lt;/load-on-startup&gt;</span>
<span>&lt;/servlet&gt;</span></code>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<h2>7. Starting up the application container</h2>
<div>
<div>
<h3>7.1. Apache Tomcat is is an easy to use servlet container</h3>
<p><a href="http://tomcat.apache.org">Apache Tomcat</a> is a servlet container. It can run JavaEE Web applications like our JAX-RS applications.</p>
<p>Tomcat Version 8 is implementing the JavaEE 7 standard supporting the latest APIs.</p>
</div>
<div>
<h3>7.2. Installing and running Apache Tomcat</h3>
<p>Installation is straight forward: After the download only few configuration steps are necessary.</p>
<p>All of these have been prepared in the script <code>tools/tomcat/run-tomcat.bat</code>. When you start it the first time, the the script will download the installation archive, unzip it and update the configuration.</p>
<div>
<h4>7.2.1. Activated Apache Tomcat features</h4>
<p>The following features are being set up using the automatic installation:</p>
<p>In <code>setenv.bat</code>:</p>
<div>
<ol>
<li>
<p>Allow Java remote debugging on port 1044<br></p>
<div>
<div>
<pre>
<code>-Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=1044</code>
</pre>
</div>
</div>
</li>
</ol>
</div>
<p>In <code>context.xml</code>:</p>
<div>
<ol>
<li>
<p>Activate <code>antiResourceLocking</code> to allow hot-deployment of applications when running Microsoft Windows as operating system</p>
</li>
</ol>
</div>
<p>In <code>tomcat-users.xml</code>:</p>
<div>
<ol>
<li>
<p>Setup an user <code>deploy</code> with password <code>deploy</code> to automatically deploy an application.</p>
</li>
<li>
<p>Setup an user <code>tomcat</code> with password <code>tomcat</code> to allow login on the management GUI.</p>
</li>
</ol>
</div>
</div>
<div>
<h4>7.2.2. Testing the Apache Tomcat installation</h4>
<p>To test the application point your browser at <a href="http://localhost:8080">http://localhost:8080</a>. You should see an Apache Tomcat start page with the welcome message (see figure <a href="#img-tomcatwelcome">Apache Tomcat Welcome Message</a>).</p>
<div>
<div><img alt="Tomcat Welcome Message" src="https://ahus1.github.io/hystrix-examples/manual/tomcatwelcome.png"></div>
<p>Figure 1. Apache Tomcat Welcome Message</p>
</div>
<p>Click on the button <strong>Manager App</strong> and you will be asked for a user name and password. Use the user <code>tomcat</code> with the password stated above to log in.</p>
</div>
</div>
</div>
</div>
<div>
<h2>8. Using runtime configuration</h2>
<div>
<div>
<h3>8.1. Runtime configuration will get you credits</h3>
<p>If you would write applications that would require a restart every time you change a setting, it will deter your users as you will have a downtime.</p>
<p>How nice if you could change all your settings at runtime! No more night shifts for IT operating to change parameters and restart the services!</p>
</div>
<div>
<h3>8.2. Netflix Archaius to the rescue</h3>
<p>Hystrix is integrated already with <a href="https://github.com/Netflix/archaius" target="_blank">Archaius</a> that handles runtime configuration for you. When implementing <code>IBANValidator</code> you&rsquo;ve seen the dynamic property being introduced:</p>
<div>
<p>IBANValidator.java</p>
<div>
<pre>
<code data-lang="java"><span>private</span> <span>static</span> DynamicLongProperty timeToWait = DynamicPropertyFactory
        .getInstance().getLongProperty(<span><span>"</span><span>hystrixdemo.sleep</span><span>"</span></span>, <span>100</span>);</code>
</pre>
</div>
</div>
<p>This will create a dynamic property. When it&rsquo;s being accessed, it will always return the most up to date value for it. When the external configuration doesn&rsquo;t define a value the default of 100 will be returned.</p>
<p>In more advanced situations you can also register a callback to be notified whenever a configuration changes. Please refer to the <a href="https://github.com/Netflix/archaius/wiki" target="_blank">Archaius Wiki</a> to find out more about it.</p>
</div>
<div>
<h3>8.3. How runtime configuration works in this project</h3>
<p>Archaius supports multiple back ends to store configuration data. In this setup all configuration is stored in a Java properties file. Archaius checks this file every second.</p>
<p>This is configured as follows:</p>
<div>
<p>setenv.bat</p>
<div>
<pre>
<code>-Darchaius.configurationSource.additionalUrls=file:///%CATALINA_HOME%/../archaius.properties
-Darchaius.fixedDelayPollingScheduler.delayMills=1000
-Darchaius.fixedDelayPollingScheduler.initialDelayMills=1000</code>
</pre>
</div>
</div>
<p>This allows you to change the dynamic:</p>
<p>This will override the default setting with 10.</p>
</div>
<div>
<h3>8.4. Configuring the Hystrix command via Archaius</h3>
<p>All Hystrix settings can be configured on default and on command level. Please refer to the <a href="https://github.com/Netflix/Hystrix/wiki/Configuration">Configuration Wiki of Hystrix</a> for an in-depth discussion. This will now show the core parameters and configure them on default level:</p>
<div>
<p>archaius.properties</p>
<div>
<pre>
<code># Should Hystrix interrupt a command that is overdue?
# default: true
hystrix.command.default.execution.isolation.thread.interruptOnTimeout=true

# How many errors are allowed before the circuit breaker is activated?
# default: 50 (must be greater than 0,
# 100 means no breaking despite of errors)
hystrix.command.default.circuitBreaker.errorThresholdPercentage=50

# How many requests are needed in the time span to trigger the circuit breaker?
# default: 20
hystrix.command.default.circuitBreaker.requestVolumeThreshold=20

# After what time (in ms) should the circuit breaker try a single request?
# default: 5000
hystrix.command.default.circuitBreaker.sleepWindowInMilliseconds=5000</code>
</pre>
</div>
</div>
<p>To give these parameters a try you should set up a load test. You&rsquo;ll see in the next chapter how you can do this. But first deploy the application to this freshly installed Tomcat!</p>
</div>
</div>
</div>
<div>
<h2>9. Building and deploying the application</h2>
<div>
<div>
<h3>9.1. Build Configuration based on conventions with Apache Maven</h3>
<p><a href="http://maven.apache.org/" target="_blank">Apache Maven</a> as a build system you the following advantages:</p>
<div>
<ol>
<li>
<p>Project layout based on conventions (i.e. your source files are in <code>src/main/java</code>)</p>
</li>
<li>
<p>Management of binary artefacts including transitive dependencies</p>
</li>
<li>
<p>A lot of plugins to automate your build</p>
</li>
</ol>
</div>
</div>
<div>
<h3>9.2. Installing Apache Maven</h3>
<p>This process has been automated for you, as all projects include <a href="https://github.com/takari/maven-wrapper" target="_blank">Maven Wrapper</a>. This will download and install maven automatically.</p>
</div>
<div>
<h3>9.3. Packaging and deploying your application to Apache Tomcat</h3>
<p>To do this run <code>hystrix-application/maven-package-deploy</code>. <strong>Please ensure that your Apache Tomcat is running,</strong> as Maven will deploy your application into the running Apache Tomcat.</p>
<p>The steps executed by Maven <code>mvn clean package cargo:redeploy</code> included in the Batch-file:</p>
<div>
<ol>
<li>
<p>Clean up the build folder <code>target</code>,</p>
</li>
<li>
<p>download all dependencies from public Maven repositories if they haven&rsquo;t been downloaded yet,<br>
(this might take a while on the first run depending on the speed of your Internet connection)</p>
</li>
<li>
<p>compile all Java sources,</p>
</li>
<li>
<p>package a WAR file in the <code>target</code> folder,</p>
</li>
<li>
<p>deploy it to your running Tomcat instance.</p>
</li>
</ol>
</div>
<p>To test that your application has been deployed successfully, open the URL <a href="http://localhost:8080/hystrixapp/api/simple" target="_blank">http://localhost:8080/hystrixapp/api/simple</a> in your browser. You should get the response <code>Hello world</code> to show you that everything has been set up correctly.</p>
</div>
</div>
</div>
<div>
<h2>10. Putting your application under load</h2>
<div>
<div>
<h3>10.1. Apache JMeter is free and open source load testing</h3>
<p><a href="http://jmeter.apache.org/" target="_blank">Apache JMeter</a> allows you to create and run load tests simulating the load of multiple users using a single PC.</p>
</div>
<div>
<h3>10.2. Installing Apache JMeter</h3>
<p>This process has been automated for you. The first time you start <code>tools/maven/run-jmeter.bat</code> JMeter is downloaded and installed.</p>
<p>The following defaults are applied:</p>
<div>
<ol>
<li>
<p>Whenever you run JMeter using the script above, the load testing script <code>hystrixSaveAccount.jmx</code> is opened.</p>
</li>
</ol>
</div>
<p>You will see the <a href="#img-jmeterwelcome">Apache JMeter Start Screen</a>.</p>
<div>
<div><img alt="JMeter Start Screen" src="https://ahus1.github.io/hystrix-examples/manual/jmeterwelcome.png"></div>
<p>Figure 2. Apache JMeter Start Screen</p>
</div>
</div>
<div>
<h3>10.3. Simulating multiple users</h3>
<p>You can vary the number of simulated users by changing the number of threads. In this script the number of users is configured to be 20.</p>
<p>A test with JMeter has the following steps:</p>
<div>
<ol>
<li>
<p>choose the number of users (aka threads)</p>
</li>
<li>
<p>press the play <span><img alt="Play" title="Play" src="https://ahus1.github.io/hystrix-examples/manual/jmeter-play.png"></span> button to start the workers</p>
</li>
<li>
<p>press the stop <span><img alt="Stop" title="Stop" src="https://ahus1.github.io/hystrix-examples/manual/jmeter-stop.png"></span> button to stop the workers</p>
</li>
<li>
<p>Watch the summary report to see the timing of the requests and how many requests succeed and fail</p>
</li>
<li>
<p>Watch the log file of Apache Tomcat for exceptions and other log messages</p>
</li>
<li>
<p>Change the Archaius runtime configuration and see how the behaviour of the system changes (see section <a href="#archaius">Using runtime configuration</a> for the details).</p>
</li>
<li>
<p>press the clean <span><img alt="Clean" title="Clean" src="https://ahus1.github.io/hystrix-examples/manual/jmeter-clean.png"></span> button to clear the results from JMeter</p>
</li>
</ol>
</div>
<div>
<table>
<tbody>
<tr>
<td></td>
<td>When you change the number of users (aka threads) it will become effective only after you stop/start the workers.</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div>
<h2>11. Aggregating Hystrix runtime information in a cluster</h2>
<div>
<div>
<h3>11.1. Turbine aggregates runtime information</h3>
<p><a href="https://github.com/Netflix/Turbine">Netflix Turbine</a> is a web application that connects to instances of your Hystrix application in a cluster and aggregates the information.</p>
<p>It does this in real time. The results will be updated every 0.5 seconds.</p>
</div>
<div>
<h3>11.2. Installing and running Turbine</h3>
<p>The installation has been scripted for you: You need to run <code>tools/turbine/maven-package-deploy.bat</code>.</p>
<p>The current installation will re-package the 0.4 version of Turbine and add to it the latest (yet unreleased) changes to allow easy re-deployment. Please have a look into <code>pom.xml</code> to see how you can adopt an already packaged WAR file to your environment.</p>
<p>You can open the URL <a href="http://localhost:8080/turbine/turbine.stream">http://localhost:8080/turbine/turbine.stream</a> in a Google Chrome browser to see the stream of events. If there are no Hystrix requests, you&rsquo;ll see only "ping" messages. You can&rsquo;t test this with Mozilla Firefox or Microsoft Internet Explorer as they will open only a file download dialogue.</p>
</div>
<div>
<h3>11.3. Configuration for Turbine</h3>
<div>
<h4>11.3.1. Configuration Turbine metadata</h4>
<p>Turbine uses the same configuration mechanism as Hystrix: Archaius. See chapter "<a href="#archaius">Using runtime configuration</a> for more information. The following configuration has already been set up for you in <code>archaius.properties</code>:</p>
<div>
<p>archaius.properties</p>
<div>
<pre>
<code>turbine.instanceUrlSuffix=:8080/hystrixapp/hystrix.stream
turbine.ConfigPropertyBasedDiscovery.default.instances=localhost
turbine.ConfigPropertyBasedDiscovery.localhost.instances=localhost</code>
</pre>
</div>
</div>
<p><code>turbine.instanceUrlSuffix</code>: This URL suffix needs to be the same for all applications in the cluster. Turbine uses this endpoint to receive a stream of runtime information.</p>
<p><code>turbine.ConfigPropertyBasedDiscovery.default.instances</code>: This lists the hosts Turbine should poll information from. In a production system there are mechanisms to discover hosts automatically. In our test scenario a list is sufficient as we have only one host.</p>
</div>
<div>
<h4>11.3.2. Preparing your application for Turbine</h4>
<p>To have your application publishing the information on the URL suffix above, you&rsquo;ll need to add a servlet to your application. This has already been packaged for you. You&rsquo;ll just need to include as a dependency of your application in <code>pom.xml</code>:</p>
<div>
<p>pom.xml</p>
<div>
<pre>
<code data-lang="xml">
<span>&lt;dependency&gt;</span>
    <span>&lt;groupId&gt;</span>com.netflix.hystrix<span>&lt;/groupId&gt;</span>
    <span>&lt;artifactId&gt;</span>hystrix-metrics-event-stream<span>&lt;/artifactId&gt;</span>
    <span>&lt;version&gt;</span>${hystrix}<span>&lt;/version&gt;</span>
<span>&lt;/dependency&gt;</span></code>
</pre>
</div>
</div>
<p>Then you register it in <code>web.xml</code> as a servlet:</p>
<div>
<p>web.xml</p>
<div>
<pre>
<code data-lang="xml">
<span>&lt;servlet&gt;</span>
        <span>&lt;description&gt;</span><span>&lt;/description&gt;</span>
        <span>&lt;display-name&gt;</span>HystrixMetricsStreamServlet<span>&lt;/display-name&gt;</span>
        <span>&lt;servlet-name&gt;</span>HystrixMetricsStreamServlet<span>&lt;/servlet-name&gt;</span>
        <span>&lt;servlet-class&gt;</span>com.netflix.hystrix.contrib.metrics.eventstream.HystrixMetricsStreamServlet<span>&lt;/servlet-class&gt;</span>
<span>&lt;/servlet&gt;</span></code>
</pre>
</div>
</div>
<p>You can open the URL <a href="http://localhost:8080/hystrixapp/hystrix.stream">http://localhost:8080/hystrixapp/hystrix.stream</a> in a Google Chrome browser to see the stream of events. If there are no Hystrix requests, you&rsquo;ll see only "ping" messages. You can&rsquo;t test this with Mozilla Firefox or Microsoft Internet Explorer as they will open only a file download dialogue.</p>
<p>To visualize it, see the next chapter about Hystrix dashboard.</p>
</div>
</div>
</div>
</div>
<div>
<h2>12. Displaying Hystrix runtime information in a browser</h2>
<div>
<div>
<h3>12.1. Hystrix Dashboard displays a Realtime Dashboard</h3>
<p><a href="https://github.com/Netflix/Hystrix/tree/master/hystrix-dashboard" target="_blank">Hystrix Dashboard</a> is part of the Hystrix distribution. It displays runtime information about every Hystrix command and thread pool for the last few minutes.</p>
</div>
</div>
</div>
<div>
<h2>13. Behaviour of Hystrix under load</h2>
<div>
<div>
<h3>13.1. Introduction to Load Scenarios for Hystrix</h3>
<p>Now we have a complete setup: an application, a driver for the load and a real time dashboard. You are now ready to see the behaviour of Hystrix.</p>
<p>The number of threads in JMeter will remain constant for all scenarios. The scenarios will differ only in their setting of <code>hystrixdemo.sleep</code> in <code>archaius.properties</code>. This will simulate an IBAN/BIC validation that is gradually getting slower and slower.</p>
<p>The delay between requests in the thread pool is 1 second. The time of the service to respond is not compensated by the delay: The longer the response time is, the fewer requests per second we will see. When there is no delay at all, we would see 20 requests per second, as we have 20 workers.</p>
</div>
<div>
<h3>13.2. A fast response: 10 ms</h3>
<p>Starting with a response time of 10 ms we see a very well behaving system: The throughput is about 18 requests per second.</p>
<p>The number of active threads is idling between one and two. The median service time is about 10 ms, the mean slightly above.</p>
<div>
<div><img alt="Hystrix Dashboard: 10 ms service delay" src="https://ahus1.github.io/hystrix-examples/manual/dashboard_10ms.png"></div>
<p>Figure 3. Hystrix Dashboard: 10 ms service delay</p>
</div>
</div>
<div>
<h3>13.3. A medium response: 70 ms</h3>
<p>Now change the delay to 70 ms. Save the file. Archaius will pick up the new value within a second. Our IBAN/BIC service will get slower.</p>
<p>You will see an immediate effect on the Dashboard. After a short time the values in the dashboard will be re-calculated.</p>
<div>
<div><img alt="Hystrix Dashboard: 70 ms service delay" src="https://ahus1.github.io/hystrix-examples/manual/dashboard_70ms.png"></div>
<p>Figure 4. Hystrix Dashboard: 70 ms service delay</p>
</div>
<p>With 13.6 requests per second and a delay of 70 ms in the service is now saturated. 13.7 req/s multiplied with 70 ms is almost 1 second working time per second.</p>
<p>You also see that the number of active threads has increased to 6.</p>
</div>
<div>
<h3>13.4. A late response: 100 ms</h3>
<p>Change the delay to 100 ms. Save the file. You&rsquo;ll see the dashboard change.</p>
<div>
<div><img alt="Hystrix Dashboard: 100 ms service delay" src="https://ahus1.github.io/hystrix-examples/manual/dashboard_100ms.png"></div>
<p>Figure 5. Hystrix Dashboard: 100 ms service delay</p>
</div>
<p>The number of active threads iterates between 9 and 10. (Please ignore the "queue size", this is not active as queueing is disabled by default). Indicated as a yellow number some requests receive a time out while they are waiting at the synchronized lock on the simulated IBAN server. Purple indicates the number of requests that are rejected when there is no space in the queue.</p>
<p>The dashboard shows you that 26 percent of all requests fail.</p>
<p>If the service wouldn&rsquo;t have been wrapped as a service, we would see an overload situation and with long response times. With Hystrix the response times in the 99 percent tier are still around one second for the successful requests.</p>
<p>The circuit is closed, meaning that Hystrix considers the service to be functional (although some requests are rejected).</p>
</div>
<div>
<h3>13.5. A critical response: 200 ms</h3>
<p>Now try 200 ms. Save the file. The dashboard will change again.</p>
<p>You&rsquo;ll see that the Circuit changes between open and closed. Therefore the following picture is a combination of two screen shots.</p>
<div>
<div><img alt="Hystrix Dashboard: 200 ms service delay" src="https://ahus1.github.io/hystrix-examples/manual/dashboard_200ms.png"></div>
<p>Figure 6. Hystrix Dashboard: 200 ms service delay</p>
</div>
<p>The ratio of failed commands is now more than 50 percent, therefore the circuit breaker kicks in. Every five seconds it tries on request (that is answered within 200 ms), resets the statistics and then closes again.</p>
<p>As the statistics are reset on each opening circuit, you can&rsquo;t trust the dashboard. Please have a look in JMeter in the Summary report. Reset it (as it will by now contain results from the previous setups) and wait a short time. It will show you that about 65 percent of the requests fail.</p>
</div>
<div>
<h3>13.6. An eternity: 1500 ms</h3>
<p>The final setting: 1500 ms. Save the file. Watch the dashboard.</p>
<div>
<div><img alt="Hystrix Dashboard: 1500 ms service delay" src="https://ahus1.github.io/hystrix-examples/manual/dashboard_1500ms.png"></div>
<p>Figure 7. Hystrix Dashboard: 1500 ms service delay</p>
</div>
<p>As the time out is 1000 ms for the commands, 100 percent of the requests fail. The circuit is now permanently open. Even the retry every 5 seconds will not close it again as this single request is too slow.</p>
</div>
<div>
<h3>13.7. Hystrix: Don&rsquo;t waste resources by queueing requests</h3>
<p>The examples show that Hystrix does its job very effectively:</p>
<div>
<ol>
<li>
<p>It will forward as many requests as possible to the service to process them.</p>
</li>
<li>
<p>Any other requests are returned to the sender as unsuccessful as early as possible.</p>
</li>
<li>
<p>No caller waits longer than the 1 second time out for a response.</p>
</li>
<li>
<p>Queues require resources and Hystrix efficiently minimizes queueing inside the system. It is limited by default to ten concurrent requests processed by the thread pool.</p>
</li>
<li>
<p>A burst in requests is reflected at the earliest boundary instead of overloading the backend service.</p>
</li>
<li>
<p>When the back end services are overloaded for one or the other reason, the circuit breaker in front will give the backend services a break to recover.</p>
</li>
</ol>
</div>
<p>Well done, Hystrix!</p>
<p>Please feel free to try your own variations: enable queueing, change the number of threads, or try out any other Hystrix setting.</p>
</div>
</div>
</div>
<div>
<h2>14. Long term monitoring of Hystrix metrics</h2>
<div>
<div>
<h3>14.1. Zabbix as classic long term monitoring and notification</h3>
<p><a href="http://zabbix.com/" target="_blank">Zabbix</a> is a classic monitoring system. It is stable and it has grown over several iterations. It supports a REST API out of the box.</p>
<p>All data is stored in a database. The server polls information from clients, runs commands or receives information from agents.</p>
<p>A frontend service will provide a web interface that includes configuration, monitoring and graphing capabilities.</p>
</div>
<div>
<h3>14.2. Installing and running Zabbix</h3>
<p>The installation has been scripted as a <a href="http://vagrantup.com" target="_blank">Vagrant</a> script. This will first download a linux image of CentOS and spin it up in <a href="http://virtualbox.org/" target="_blank">Virtualbox</a>. Then it will hand over to a provisioner, in our case this is <a href="http://saltstack.com/" target="_blank">Saltstack</a>.</p>
<p>To install Vagrant please see <a href="#vagrant">Running virtual machines for test and development</a>. Once it is installed, run the follwing command in <code>tools/zabbix</code>:</p>
<p>This will take a while to download all the files necessary. Once it is up and running point your browser at your local Zabbix installation: <a href="http://localhost:2280/zabbix">http://localhost:2280/zabbix</a>.</p>
<p>Use Username <code>Admin</code> password <code>zabbix</code> to log in (please note the capital <strong>A</strong> in the user name)</p>
</div>
<div>
<h3>14.3. Configuration of Zabbix</h3>
<p>All configuration items could be automated using the REST API. This will be a future task for this tutorial.</p>
<p>For now these are the manual steps:</p>
<div>
<ol>
<li>
<p>In click on <b>Import</b>. Pick the file <code>hystrix_template.xml</code> from <code>tools\zabbix</code>. Click on <b>Import</b>.</p>
</li>
<li>
<p>Choose again. You see a <code>hystrixTemplate</code> here. Click on this template to edit it. Include the host <code>hystrix</code> in group <code>Discovered Hosts</code> in the shuttle view to put it on the left side. This will apply the template to the host. Click on <b>Save</b>.</p>
</li>
<li>
<p>In click on <b>Create Host</b>. Enter <code>hystrix</code> as host name here. Pick <code>Discovered Hosts</code> as the Host group. Now <b>Save</b>.</p>
</li>
</ol>
</div>
<p>This is all configuration that is needed: You have created a host, and you have associated it with a template that will autodiscover all Hystrix comamnds.</p>
</div>
<div>
<h3>14.4. Configuration of the example application</h3>
<p>As default the Zabbix agent in the application is disabled. To enable it, go to the file <code>archaius.properties</code>. Change the <code>hystrixdemo.enablezabbix</code> to be true.</p>
<p>The change is active at runtime. This will trigger the activation in <code>HystrixSetupListener</code>.</p>
</div>
<div>
<h3>14.5. Zabbix autodiscovery at work</h3>
<p>The following steps will now happen automatically:</p>
<div>
<ol>
<li>
<p>The Zabbix agent will connect to <code>127.0.0.1:10051</code> and identifies itself as host <code>hystrix</code>. In our example this is hardcoded.</p>
</li>
<li>
<p>The Zabbix agent asks the server for any scheduled checks. The only check that is configured for the host <code>hystrix</code> is the <code>hystrixCommand.discovery</code> in the template.</p>
</li>
<li>
<p>The agent runs the check and delivers all active Hystrix commands as a JSON object to the server.</p>
</li>
<li>
<p>This will trigger the auto discovery. For every command Zabbix will create 20 items and three graphs.</p>
</li>
<li>
<p>When the Zabbix agent reloads the list of checks after a minute, there will be a lot more checks. It will run them as specified in the interval of the template we imported to Zabbix.</p>
</li>
</ol>
</div>
<p>The auto discovery rule is set to run every five seconds. Please note that the auto discovery will only find commands that have been triggered at least once during the runtime of the application. Hystrix will only notify our Zabbix agent at the first created command.</p>
<p>Please choose and you should see the number of Items change for the host <code>hystrix</code>. Please refresh your browser if necessary.</p>
<p>Now choose and choose the host <code>hystrix</code> and any of the three graphs. After a few minutes you should see the first graphs appearing. The following screen shows a service that was answering successfully first. After a short time the response time was changed to 250 ms. Soon lots of rejected and timeout requests appear.</p>
<div>
<div><img alt="Zabbix counting Hystrix requests" src="https://ahus1.github.io/hystrix-examples/manual/requestcount_zabbix.png"></div>
<p>Figure 8. Zabbix counting Hystrix requests</p>
</div>
<p>Zabbix polls for the command data every minute. Change the timing in <code>archaius.properties</code> to see information about failed requests.</p>
<p>Click on to see the latest values that the agent has sent.</p>
</div>
<div>
<h3>14.6. Java classes to forward information from Hystrix to Zabbix</h3>
<p>In order to activate the forwarding of events, you&rsquo;ll need to change <code>hystrixdemo.enablezabbix</code> to <code>true</code> in <code>archaius.properties</code>. This change will be active immediately.</p>
<p>The property <code>hystrixdemo.enablezabbix</code> in <code>archaius.properties</code> is evaluated in the class <code>HystrixSetupListener</code>. When this property is enabled, the Zabbix agent is started.</p>
<p>You find all additional classes in the package <code>de.ahus1.hystrix.util.zabbix</code> in the example application. The following classes work together to forward the information to Zabbix:</p>
<div>
<ol>
<li>
<p><code>ZabbixCommandMetricsProvider</code> is registered with Zabbix. It handles the connection to Zabbix. It creates a callback <code>ZabbixCommandMetricsProvider</code> that will be called by the agent when it requires data for the provider <code>hystrixCommand</code>. It creates for Zabbix on request a <code>HystrixZabbixMetricsPublisherCommand</code> when a command is created for the first time.</p>
</li>
<li>
<p>Once the <code>HystrixZabbixMetricsPublisherCommand</code> initialized, it registers itself with the <code>ZabbixCommandMetricsProvider</code>.</p>
</li>
<li>
<p>The <code>ZabbixCommandMetricsProvider</code> will provide a list of HystrixCommand keys when it is called with <code>hystrixCommand.discovery</code> When it is called <code>hystrixCommand.countSuccess[{#COMMAND}]</code> it returns the current count of successful Hystrix command calls.</p>
</li>
</ol>
</div>
<p>The current implementation doesn&rsquo;t return any information about thread pools yet.</p>
</div>
<div>
<h3>14.7. Working with templates in Zabbix</h3>
<p>When you change your template, the changes are applied to all hosts assigned to the template. This makes it easy to ensure a common configuration of all Hystrix commands throughout your infrastructure.</p>
<p>You can also choose to setup triggers automatically.</p>
<p>When a command is no longer used in your application, the auto discovery will no longer return it to Hystrix. Hystrix will remove all items that are no longer needed after 30 days. You can simulate this behaviour by stopping/starting the tomcat server and not sending any requests via JMeter. As now commands will be triggered, the list of auto-discovered commands is empty. A few seconds later Zabbix will add a yellow exclamation mark to your items and will start the 30 day countdown.</p>
</div>
</div>
</div>
<div>
<h2>15. Real time monitoring of Hystrix metrics</h2>
<div>
<div>
<h3>15.1. Riemann as cutting edge realtime event processing</h3>
<p><a href="http://www.riemann.io/" target="_blank">Riemann</a> is a a modern event management system. It has been designed with with performance in mind for montoring complex systems. To process the data fast it keeps the last few minutes of data in memory. All events are processed in a streaming mode.</p>
<p>The events can be aggregated and filtered. You can also start actions like notifying IT operations of problems.</p>
<p>Riemann comes with its own dashboard that can show you can use to show real time graphs and statistics.</p>
<p>So far both the Hystrix Dashboard and Zabbix have been presented. You could consider Riemann for the following:</p>
<div>
<ol>
<li>
<p>Zabbix polls existing metrics from Hystrix. You can aggregate events to deliver your own metrics using Riemann. This all happens outside of the application. This gives you a lot of flexibility.</p>
</li>
<li>
<p>While the Hystrix Dashboard is a great start for a dashboard, it is not configurable. The Riemann Dashboard is highly configurable (once you understand its keyboard shortcuts) and might delivery you additional insights you might need to administer your cluster.</p>
</li>
<li>
<p>Assuming that Zabbix receives information every minute, this means that you will receive alerts with a delay of one minute. After a problem disappeared (for example after you have restarted a service, or fixed a configuration problem) you will see this only after a minute or two in the Zabbix graphs. Riemann gives you the options to be notified immediately by a channel of your choice (email, SMS, chat) when the problem occurs and when it is fixed.</p>
</li>
</ol>
</div>
<p>Filtered events can be forwarded to systems like InfluxDB to be stored. These can be displayed later using tools like Grafana. But lets start with Riemann first.</p>
</div>
<div>
<h3>15.2. Installing and running Riemann</h3>
<p>The installation has been scripted as a <a href="http://vagrantup.com" target="_blank">Vagrant</a> script. This will first download a linux image of CentOS and spin it up in <a href="http://virtualbox.org/" target="_blank">Virtualbox</a>. Then it will hand over to a provisioner, in our case this is <a href="http://saltstack.com/" target="_blank">Saltstack</a>.</p>
<p>To install Vagrant please see <a href="#vagrant">Running virtual machines for test and development</a>. Once it is installed, run the follwing command in <code>tools/riemann</code>:</p>
<p>This will take a while to download all the files necessary. Once it is up and running point your browser at your local Riemann installation: <a href="http://localhost:4567/">http://localhost:4567/</a>.</p>
</div>
<div>
<h3>15.3. How Events are captures for Riemann</h3>
<p>In the example application the <code>HystrixSetupListener</code> registers the class <code>HystrixRiemannEventNotifier</code> with Hystrix. This will be notified with the detailed timinings of each run command and with the outcome of each command (successful, timeout, etc.).</p>
<p>These events are queued in a bounded queue. The result will be forwarded to Riemann in batches every 100 ms.</p>
<p>In order to activate the forwarding of events, you&rsquo;ll need to change <code>hystrixdemo.enableriemann</code> to <code>true</code> in <code>archaius.properties</code>. This change will be active immediately.</p>
</div>
<div>
<h3>15.4. Configuring Riemann Server</h3>
<p>In this setup the information about Hystrix commands is forwarded on a level of a single command execution.</p>
<p>We can present it on this level, but usually you want to create aggreation on top of the event stream. The first listing aggregates the timings of the service <code>IBANValidatorCommand</code>.</p>
<div>
<p>Percentiles with riemann.config</p>
<div>
<pre>
<code data-lang="closure">(let [index (index)]
  (streams
    (where (and (= service "IBANValidatorCommand") (not (= metric nil)) )

       ; Now, we'll calculate the 50th, 95th, and 99th percentile for all
       ; requests in each 5-second interval.

       (percentiles 5 [0.5 0.95 0.99]

           ; Percentiles will emit events like
           ; {:service "api req 0.5" :metric 0.12}
                   ; We'll add them to the index, so they can show up
           ; on our dashboard.

           index)
    )
  )
)</code>
</pre>
</div>
</div>
<p>The second listing counts the different outcomes (SUCCESS, TIMEOUT, etc.) in every five second interval. This statement contains a statement for logging the aggregated events in the Riemann server log.</p>
<div>
<p>Counting with riemann.config</p>
<div>
<pre>
<code data-lang="closure">(let [index (index)]
  (streams
    (where (and (= service "IBANValidatorCommand") (= metric nil) )

      (by :state
        (with :metric 1 (with :service "IBAN Count"
          (rate 5
          ; log all events (for debugging)
          #(info "received event" %)
            index
          )
        ))
      )

    )
  )
)</code>
</pre>
</div>
</div>
</div>
<div>
<h3>15.5. Configuring Riemann Dashboard</h3>
<p>You can configure the frontend of Riemann in your browser. This setup installs a dashboard that shows some useful information about the example application.</p>
<div>
<div><img alt="Riemann Dashboard" src="https://ahus1.github.io/hystrix-examples/manual/riemann_dashboard.png"></div>
<p>Figure 9. Riemann Dashboard</p>
</div>
<p>In the lower part of the screen you will see a guide how to use keyboard shortcuts to customize the dashboard.</p>
<p>As with the Zabbix monitoring, this will only show useful information once some commands are run. Please use the JMeter load test for this.</p>
</div>
</div>
</div>
<div>
<h2>16. Writing an online manual</h2>
<div>
<div>
<h3>16.1. Asciidoctor is a lean tool for technical documentation</h3>
<p>This manual is written in Asciidoc format. Asciidoctor is used to create an online documentation from it.</p>
<p>To find out more about Asciidoctor see the <a href="http://asciidoctor.org/">http://asciidoctor.org/</a>. It is a clone of the original Asciidoc using Ruby. Compiled versions are available for Java and JavaScript. This allows you to easily integrate it into your build process.</p>
<p>You can create from a single source of documentation formats like HTML5, PDF and epub.</p>
</div>
<div>
<h3>16.2. Editors to work on Asciidoctor</h3>
<p>Asciidoctor files are plain text files with an easy to learn syntax. You can use any text editor to write the files.</p>
<p>You might want to try the <a href="https://github.com/asciidoctor/atom-asciidoc-preview" target="_blank">Asciidoctor preview plugin</a> and <a href="https://atom.io/packages/language-asciidoc" target="_blank">Asciidoctor language plugin</a> for <a href="https://atom.io/" target="_blank">GitHub&rsquo;s Atom</a> editor. As the time of writing this provides you the best authoring experience.</p>
</div>
</div>
</div>
<div>
<h2>17. Running virtual machines for test and development</h2>
<div>
<div>
<h3>17.1. Vagrant manages virtual machines</h3>
<p>Virtual machines make development and test easier, as they can be started and stoped as you need them. You re-use the CPU, RAM and hard disk space on your PC. When you don&rsquo;t need them you shut them down and they will use only some hard drive space. Once your test or development are over you can wipe them from your hard disk.</p>
<p>A software to provide this virtualization is the free <a href="http://virtualbox.org/" target="_blank">Oracle&rsquo;s Virtualbox</a>. You can spin up as many machines as you have RAM, CPU and hard disk space on your local machine. You can set up custom networking and interact with the machines either via screen or network.</p>
<p>But setting up a test environment would still be a manual process if you would use only Virtualbox. <a href="http://vagrantup.com" target="_blank">Vagrant</a> builds on top of it and provides a command line and scripting for you. It also allows you to create local test instances from centrally managed ready made boxes, so you don&rsquo;t need to install an operating system manually.</p>
</div>
<div>
<h3>17.2. Installing Vagrant and useful plugins</h3>
<p>To use Vagrant please use the following steps:</p>
<div>
<ol>
<li>
<p>Download and install Oracle&rsquo;s Virtualbox from <a href="http://virtualbox.org/">http://virtualbox.org/</a>.</p>
</li>
<li>
<p>Download and install Hashicorp&rsquo;s Vagrant from <a href="http://vagrantup.com">http://vagrantup.com</a>.</p>
</li>
<li>
<p>Install the Vagrant Cachier Plugin (<a href="https://github.com/fgrehm/vagrant-cachier">https://github.com/fgrehm/vagrant-cachier</a>) to cache all downloaded software packages locally. This will speed up future installations. Do this by starting from the command line:</p>
<div>
<div>
<pre>
vagrant plugin install vagrant-cachier
</pre>
</div>
</div>
</li>
<li>
<p>Install the Vagrant Guest Additions Plugin (<a href="https://github.com/dotless-de/vagrant-vbguest">https://github.com/dotless-de/vagrant-vbguest</a>) to be able to update the virtual box tools on the downloaded images automatically. Do this by starting from the command line:</p>
<div>
<div>
<pre>
vagrant plugin install vagrant-vbguest
</pre>
</div>
</div>
</li>
<li>
<p>Install the Vagrant proxy plugin (<a href="https://github.com/tmatilai/vagrant-proxyconf">https://github.com/tmatilai/vagrant-proxyconf</a>) to configure the downloaded image automatically regarding its proxy settings. Do this by starting from the command line</p>
<div>
<div>
<pre>
vagrant plugin install vagrant-proxyconf
</pre>
</div>
</div>
</li>
</ol>
</div>
</div>
<div>
<h3>17.3. Proxy configuration for Vagrant</h3>
<p>To be able to install any vagrant plugin behind a proxy, set the <code>http_proxy</code> environment variable. The this syntax assumes you are running a Microsoft Windows system:</p>
<div>
<div>
<pre>
SET HTTP_PROXY="http://username:password@proxyhost:proxyport/
SET HTTPS_PROXY="http://username:password@proxyhost:proxyport/
</pre>
</div>
</div>
<p>Vagrant will not re-use your system&rsquo;s proxy settings. You&rsquo;ll need to specify them again.</p>
<p>I recommend that you set up a common configuration for all your boxes (not individually for each box). Create (or ammend) the file <code>Vagrantfile</code> in your <code>%USERPROFILE%\.vagrant.d</code> folder.</p>
<div>
<p>Vagrantfile</p>
<div>
<pre>
<code># URI of the local (caching) HTTP proxy
# LOCAL_HTTP_PROXY = 'http://proxyhost:proxyport'
LOCAL_HTTP_PROXY =

# Configures vagrant-cachier and vagrant-proxyconf.
# Should be called only on "local machine" providers.
def configure_caching(config)
  if Vagrant.has_plugin?('vagrant-cachier')
    config.cache.enable :gem
    config.cache.enable :npm
  end

  if Vagrant.has_plugin?('vagrant-proxyconf')
    config.proxy.http     = LOCAL_HTTP_PROXY
    config.proxy.https    = LOCAL_HTTP_PROXY
    config.proxy.no_proxy = 'localhost,127.0.0.1'
  end
end

Vagrant.configure('2') do |config|
  config.vm.provider :virtualbox do |vbox, override|
    configure_caching(override)
  end
end</code>
</pre>
</div>
</div>
<p>Alternatively you can set the environment variables <code>VAGRANT_HTTP_PROXY</code>, <code>VAGRANT_HTTPS_PROXY</code> and <code>VAGRANT_NO_PROXY</code>.</p>
</div>
<div>
<h3>17.4. Managing virtual machines from the command line</h3>
<p>You can start a box with:</p>
<p>You can stop a box with:</p>
<p>You can log in to the box using:</p>
<p>For a list of other commands start</p>
<p>By default a user <code>vagrant</code> with password <code>vagrant</code> exists every machine. You can use <code>ssh</code> to access a console. From the <code>vagrant</code> user you can use the <code>sudo</code> command to acquire root access.</p>
</div>
</div>
</div>
<div>
<h2>18. Automatic provisioning of machines</h2>
<div>
<div>
<h3>18.1. Saltstack installation automation</h3>
<p>In order to provide reproducable and easy to setup development and test environments, you can script the installation of the software.</p>
<p>There are different tools available to install and configure software once the base operating system is installed. Some of the tools used for this are Chef, Puppet, Saltstack and Ansible.</p>
<p>I&rsquo;ve chosen Saltstack here as it concentrates on its task to install and configure software - nothing more. For this it takes input as simple YAML files with only very little scripting.</p>
<p>Vagrant already knowns about different provisioners, also about Saltstack. To hand over control to Saltstack only the following few lines are necessary in the <code>Vagrantfile</code> of the virtual machine you want to provision:</p>
<div>
<p>Vagrantfile</p>
<div>
<pre>
<code>config.vm.provision :salt do |salt|
  salt.minion_config = "salt/minion"
  salt.run_highstate = true
end</code>
</pre>
</div>
</div>
<p>Vagrant will first check if Saltstack is already installed on the virtual machine. If not, it will install the latest version of Saltstack. Once this is complete it hands over control to Saltstack to install all the configured packages.</p>
<p>To find out more about Saltstack please visit its website <a href="http://saltstack.com">http://saltstack.com</a>.</p>
</div>
<div>
<h3>18.2. Running Saltstack in standalone mode</h3>
<p>Saltstack can be run in two modes:</p>
<div>
<ol>
<li>
<p><strong>Centrally managed configuration</strong>: a master server stores all configuration and passes it to the clients (the clients are called "minions").</p>
</li>
<li>
<p><strong>Locally stored configuration</strong>: A client will have all configuration files necessary stored locally. This setup is called "masterless" setup.</p>
</li>
</ol>
</div>
<p>The scripts will not differ in the two setups. But it simlifies development and also the setup we use here to have a masterless setup. To do this, I specied <code>file_client: local</code> in the <code>minion</code> file.</p>
<p>(Note: this masterless setup is not unique to Saltstack, it is also available for other provisioning tools.)</p>
</div>
<div>
<h3>18.3. Configuration files for automatic installation</h3>
<p>You will find the configuration files in <code>salt/roots/salt</code>. They are plain text files. You can open them in a plain text editor to get a first impression. Please refer to the Salstack documentation find out more.</p>
</div>
<div>
<h3>18.4. Saltstack commands</h3>
<p>Provisioning via saltstack is run automatically. To run salt stack provisioning manually again (for example after you have changed a configuration file) use the following command as user <code>root</code>:</p>
<div>
<div>
<pre>
salt-call --local state.highstate
</pre>
</div>
</div>
<p>Please refer to the Salstack documentation find out more.</p>
</div>
</div>
</div>
<div>
<h2>19. Building the documentation automatically</h2>
<div>
<div>
<h3>19.1. Things that need maintenance</h3>
<p>This manual is written in Asciidoc format. Asciidoctor is used to create an online documentation from it.</p>
<p>This can be done manually by calling <code>manual/build.bat</code>. The changes then need to be published to the <code>gh-pages</code> branch.</p>
<p>To find out more about Asciidoc and Asciidoctor see see section <a href="#asciidoctor">Writing an online manual</a>. To find out more about GitHub Pages see <a href="https://pages.github.com/">https://pages.github.com/</a>.</p>
</div>
<div>
<h3>19.2. Travis CI as a service in the cloud</h3>
<p>Travis CI is a build service in the clould that is free to all open source projects. It integrates well with GitHub. A new build is triggered on every push to your repository.</p>
<p>The following bit&rsquo;s and pieces are needed:</p>
<div>
<ol>
<li>
<p>Hook up <a href="http://travis-ci.org/">http://travis-ci.org/</a> to your repository.</p>
</li>
<li>
<p>Add a file <code>.travis.yml</code> to the root of the repository. In our case this indicates to Travis that we want to use Ruby.</p>
</li>
<li>
<p>Add a file <code>Gemfile</code> to the root of the repository to specify the dependencies of Asciidoctor.</p>
</li>
<li>
<p>Add a file <code>manual/build.sh</code> to the project to copy images in the proper place, and run asciidoctor.</p>
</li>
<li>
<p>Add a file <code>manual/update-gh-pages.sh</code> to store the created Asciidoctor output to the repositories GitHub pages branch.</p>
</li>
</ol>
</div>
<p>You can see the current status of the build here: <a href="https://travis-ci.org/ahus1/hystrix-examples">https://travis-ci.org/ahus1/hystrix-examples</a>.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via ahus1.github.io https://ahus1.github.io<br>
<br>
January 2, 2019 at 02:46PM
